### PR TITLE
Dev guide for WHERE predicate pushdown

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by MiguelWeezardo** [comment]
Original PR: https://github.com/nineinchnick/trino/pull/9

File: docs/src/main/sphinx/develop/connectors.rst [hunk 1]
```diff
@@ -225,6 +225,205 @@ a :ref:`connector-page-sink-provider`.
...
+                ...);
+
+        return Optional.of(
+                remainingExpression.isPresent()
+                        ? new ConstraintApplicationResult<>(handle, remainingFilter, remainingExpression.get(), precalculateStatisticsForPushdown)
+                        : new ConstraintApplicationResult<>(handle, remainingFilter, precalculateStatisticsForPushdown));
+    }
+
+In this example we are implementing a base class for many JDBC connectors,
+so we have to handle all the nuances of different JDBC-capable databases.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:411]: If this doesn't explain all the nuances, only the one example below, it's not relevant. Can you rephrase this to be more explicit about having to ensure that if a constraint gets pushed down, it should work exactly the same in the underlying data source and produce the same results?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 2]
```diff
@@ -225,6 +225,205 @@ a :ref:`connector-page-sink-provider`.
...
+For example, string comparison in some databases is case-insensitive, which
+means for those databases we can't push it down since it would change it's
+meaning.
+
+To handle push down for all the different types we need to ask our JDBC client
+for a ``ColumnMapping`` which can produce a ``PredicatePushdownController``
+which decides if a column domain can be pushed down.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:418]: This is confusing. I'm not sure how I'd use this information when writing a non-JDBC connector. Note that I think we should have a whole separate page about JDBC connectors.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 3]
```diff
@@ -225,6 +225,205 @@ a :ref:`connector-page-sink-provider`.
...
+
+Once the ``TupleDomains`` are handled, expressions can be pushed down as well.
+We split the constraint expression into conjuncts - smaller expressions joined
+with AND - and handle them individually. Each one is converted using
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:422]: This is too passive, who handles the `TupleDomains`? Where do the expressions get split? Does this need to happen in the connector?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 4]
```diff
@@ -250,6 +250,10 @@ a given query.
 
 Otherwise, these methods return a result object containing a new table handle.
 
+Each of these methods receives the table handle created thus far, and can add
+additional push down information to it by returning a new table handle created
+from the old one.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:255]: This paragraph duplicates the next one and the previous one combined. What are you trying to say here that's not already in the surrounding paragraphs?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 5]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
 
+.. _predicate-pushdown:
+
+Predicate pushdown
+^^^^^^^^^^^^^^^^^^
+
+When executing a query with a ``WHERE`` clause, the query plan may
+contain a ``Filter`` operation.
+
+When the query plan contains a ``Filter`` operation, the Trino engine
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:312]:
```suggestion
When the query plan contains a ``Filter`` operation, Trino
```

File: docs/src/main/sphinx/develop/connectors.rst [hunk 6]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+tries to optimize the query by pushing down the predicate constraint
+into the connector by calling the ``applyFilter`` method of the
+connector metadata service. This method receives a table handle with
+all optimizations applied thus far, and should return either
+``Optional.empty()`` or a response with a new table handle derived from
+the old one. The constraint which will eventually be applied on the
+underlying source level is accumulated over multiple calls in a
+``ConnectorTableHandle``.  Once the query actually runs,
+``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``
+use whatever optimizations were pushed down to ``ConnectorTableHandle``.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:322]: This is true for all `applyXXX` functions, so maybe it should be added earlier, somewhere around line 250?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 7]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+The query optimizer may call ``applyFilter`` for a single query multiple times,
+as it searches for an optimal query plan. It's important that connectors
+return ``Optional.empty()`` from ``applyFilter`` if they can't apply the
+constraint for this invocation even if they support ``WHERE`` clause pushdown
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:327]: Connectors push down the `Filter` operation, not the `WHERE` clause.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 8]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+in general. It's also important to return ``Optional.empty()`` if the
+constraint has already been applied by some previous invocation.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:329]:
```suggestion
constraint has already been applied.
```

File: docs/src/main/sphinx/develop/connectors.rst [hunk 9]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+A ``Domain`` is either a list of possible values or a list of ranges.
+It also contains information about nullability.
+
+If both a predicate and a summary are available, handling the predicate
+is preferred over handling the summary, as the predicate is guaranteed
+to be more strict in filtering out values.
+However it's not possible to store a predicate in the table handle and use
+it later, as the predicate cannot be held on to after the ``applyFilter``
+call returns. This means a predicate can only be used for immediate
+filtering of entire Hive partitions/DB shards and is not actually pushed down.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:352]: This guide should not refer to any specific connectors. What does `immediate filtering` mean? `applyFilter` doesn't handle any data, other services do, like the split manager and record set provider (or page source provider).

File: docs/src/main/sphinx/develop/connectors.rst [hunk 10]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+This overlap between the predicate and summary is due to historical reasons,
+as simple comparison pushdown was implemented first via summary, and more
+complex filters such as ``LIKE`` which required more expressive predicates
+were added later.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:357]: How is that important when developing a new connector? If a connector could only handle either the summary or predicates, can this be stated explicitly?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 11]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+If a constraint can only be pushed down partially, e.g. a connector for
+a database which doesn't support range matching receives a query with
+``WHERE x = 2 AND y > 5``, the ``y`` column constraint should be
+returned in the ``ConstraintApplicationResult`` from ``applyFilter``.
+In this case the ``y > 5`` part of the filter will be applied by Trino
+engine, and not pushed down.
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:364]:
```suggestion
In this case the ``y > 5`` part of the filter will be applied Trino,
and not pushed down.
```

File: docs/src/main/sphinx/develop/connectors.rst [hunk 12]
```diff
@@ -297,6 +301,233 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+call returns. This means a predicate can only be used for immediate
+filtering of entire Hive partitions/DB shards and is not actually pushed down.
+
+This overlap between the predicate and summary is due to historical reasons,
+as simple comparison pushdown was implemented first via summary, and more
+complex filters such as ``LIKE`` which required more expressive predicates
+were added later.
+
+If a constraint can only be pushed down partially, e.g. a connector for
+a database which doesn't support range matching receives a query with
```
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:360]: Connectors don't receive queries. They are used in catalogs which can be referenced in queries.
```suggestion
If a constraint can only be pushed down partially, e.g. a connector for
a database that doesn't support range matching is used in a query with
```

File: docs/src/main/sphinx/develop/connectors.rst (hunk 9 above)
Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:352]: By `immediate filtering` I meant filtering that occurs inside `applyFilter`, since we can't store the predicate in a `TableHandle`. How this would be done depends on the DB implementation - in Hive connector we query the HMS for partitions, filter them using the predicate, and store the resulting list in `HiveTableHandle`. I guess sharded DBs could use a similar approach if their API allows querying for a list of shard keys.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:345]: This is an implementation detail that we shouldn't include in the guide. I'm not entirely sure if `applyFilter` should be doing time-consuming operations like talking to external services, since query planning should be done as quickly as possible. Instead, maybe it's better to do it in the split manager, using the table handle returned here.

It's enough to say that predicate objects cannot be reused outside of this method, we don't need to go into details about how to deal with it.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:350]: The API ambiguity should be explained by git commits or in the API Javadocs.

Regarding the preference, it should be possible only to handle one kind of constraint, either the summary or predicate, not both. So instead of writing (pseudo) code like:
```
if (predicate) {
 // handle predicate
} else {
 // handle summary
}
```
I'd rather write:
```
if (summary) {
  // handle
}
```
Are there cases when a predicate is present but the summary is not? Is that a bug?

Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:350]: I think there can be cases where the summary is present, but there is no predicate, but not the other way around.

Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:350]: Also, handling of the predicate is entirely optional - it's correct (but less efficient) to ignore it and focus on summary.

**Comment by mosabua** [comment]
There is a lot of feedback ... let me know when that is addressed and then I can look... sounds good?

In general.. shorter code snippets and less of them directly in the docs is probably better ..

**Comment by MiguelWeezardo** [comment]
@mosabua You can take a look now, thanks!

File: docs/src/main/sphinx/develop/connectors.rst [hunk 13]
```diff
@@ -297,6 +298,232 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+Once we split all ``TupleDomains`` into those that can be pushed down, and
+those which will be handled in Trino, expressions can be separated as well.
+We start by splitting each constraint expression into conjuncts - smaller
+expressions joined with AND - and handle each conjunct individually. Each one
+is converted using connector-specific rules, as defined by our ``JdbcClient``
+implementation to be more flexible. This provides an opportunity for each
+JDBC-based connector to use the full expressive power of the underlying database
+language. Those conjuncts which weren't converted are returned as
+``remainingExpression`` and will be evaluated by the Trino engine.
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/connectors.rst:525]:
```suggestion
``remainingExpression`` and are evaluated by the Trino engine.
```

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/develop/connectors.rst [hunk 14]
```diff
@@ -297,6 +298,232 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
 error code and an actionable message, to let users know how to write a valid
 query.
 
+.. _dev-predicate-pushdown:
+
+Predicate pushdown
+^^^^^^^^^^^^^^^^^^
+
+When executing a query with a ``WHERE`` clause, the query plan may
```
[the remark this answers is not in the document]
Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:306]: It's always possible for the `ScanFilterProject` operation to be optimized out at a higher level, so I think the best we can do is "will likely contain".

**Comment by MiguelWeezardo** [comment]
@mosabua Could you take another look?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 15]
```diff
@@ -303,8 +303,14 @@ query.
...
 ^^^^^^^^^^^^^^^^^^
 
-When executing a query with a ``WHERE`` clause, the query plan may
-contain a ``ScanFilterProject`` operation.
+When executing a query with a ``WHERE`` clause, the query plan will likely
+contain a ``ScanFilterProject`` operation with a predicate constraint.
+
+A predicate constraint is a description of the constraint imposed on the
+results of the query as expressed in the ``WHERE`` clause. For example
+``WHERE x > 5 AND y = 3`` would translate into a ``Constraint`` where the
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:311]: No use of "would"

...= 3`` translates into

File: docs/src/main/sphinx/develop/connectors.rst [hunk 16]
```diff
@@ -303,8 +303,14 @@ query.
...
+``summary`` field would say the ``x`` column's domain has to be greater than
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:312]: - No use of "would"
- Do not use "say" - google style guide requires plain, non-idiomatic language, and inanimate objects don't actually say things. 
 
- You can use "means" instead of "would say".

Use "must be" instead of "has to be"

File: docs/src/main/sphinx/develop/connectors.rst [hunk 17]
```diff
@@ -352,19 +359,22 @@ as simple comparison pushdown was implemented first via summary, and more
 complex filters such as ``LIKE`` which required more expressive predicates
 were added later.
 
-If a constraint can only be pushed down partially, e.g. a connector for
+If a constraint can only be pushed down partially, for example a connector for
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:362]: for example when a

File: docs/src/main/sphinx/develop/connectors.rst [hunk 18]
```diff
@@ -352,19 +359,22 @@ as simple comparison pushdown was implemented first via summary, and more
...
 a database that doesn't support range matching is used in a query with
 ``WHERE x = 2 AND y > 5``, the ``y`` column constraint should be
 returned in the ``ConstraintApplicationResult`` from ``applyFilter``.
-In this case the ``y > 5`` part of the filter will be applied in Trino,
+In this case the ``y > 5`` part of the filter is applied in Trino,
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:366]: In this case, the ``y > 5`` condition is

("part of the filter" is very informal. Please be specific about that part of the filter. It is a condition.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 19]
```diff
@@ -393,13 +403,109 @@ and our ``ConstraintApplicationResult`` can set ``remainingFilter`` to
 This pushdown implementation is quite similar to many Trino connectors
 - see ``MongoMetadata``, ``BigQueryMetadata``, ``KafkaMetadata``.
 
-Let's move on to a more complex example with expression pushdown (a simplified
+Let's move on to a more complex example where the data types from Trino are not
+exactly represented in the underlying database (a simplified
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:407]: Far too informal. This is blog post/user guide language, not reference material language. the material in the parenthesis does not make sense and in general, parenthesis should be avoided at all costs.. "Not exactly represented" is confusing phrasing.

Suggest:

The following, more complex example shows data types from Trino that are available directly in the underlying data source, and must be mapped:

File: docs/src/main/sphinx/develop/connectors.rst [hunk 20]
```diff
@@ -393,13 +403,109 @@ and our ``ConstraintApplicationResult`` can set ``remainingFilter`` to
...
+                newDomain,
+                ...);
+
+        return Optional.of(
+                new ConstraintApplicationResult<>(
+                    handle,
+                    remainingFilter));
+    }
+
+In this example we are implementing a base class for many JDBC connectors,
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:481]: Writing is too informal/blog-posty.

This example illustrates implementing a base class for many JDBC connectors while handling the specific requirements of multiple JDBC-compliant data sources. It ensure that if a constraint gets pushed down, it works
exactly the same in the underlying data source, and produce the same results
as it would in Trino.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 21]
```diff
@@ -393,13 +403,109 @@ and our ``ConstraintApplicationResult`` can set ``remainingFilter`` to
...
+so we have to handle all the nuances of different JDBC-capable databases.
+We need to ensure that if a constraint gets pushed down, it should work
+exactly the same in the underlying data source and produce the same results
+as it would in Trino.
+
+For example, string comparison in some databases is case-insensitive, which
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:487]: Move this to previous paragraph. 

Do not use "we" - too informal.

For example, in databases where string comparisons are is case-insensitive, pushdown does not work, as string
comparison operations in Trino are case-sensitive.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 22]
```diff
@@ -393,13 +403,109 @@ and our ``ConstraintApplicationResult`` can set ``remainingFilter`` to
...
+To handle push down for all the different JDBC types we need a
+``PredicatePushdownController`` which decides if a column domain
+can be pushed down. In the example code above we obtain it from a
+``JdbcClient`` implementation specific to that database. Note that
+this interface is only needed because we are trying to be generic. In non-JDBC
+databases the rules for pushing down different types would usually be
+implemented directly, without going through a ``PredicatePushdownController``
+interface.
+
+Now let's add expression pushdown to our code, enabled by a session flag:
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:500]: Do not use "we" - too informal.
Do not use "usually" (or "could"  or "would") as per google dev style guide. In reference material, writing must be definite.
Do not use directions such as above, below. User following/preceding (g dev style guide)

Also, I have used "interface" below, because you seem to call it an interface in the paragraph. Please ensure that it is being named *precisely* what it is (interface/class/method, etc).

The ``PredicatePushdownController`` interface determines if a column domain can be pushed down in JDBC-compliant data sources. In the preceding example, it is called from a ``JdbcClient`` implementation specific to that database. In non-JDBC-compliant data sources, type-based push downs are implemented directly, without going through the ``PredicatePushdownController`` interface.

Now let's add expression pushdown to our code, enabled by a session flag:

**Review by m57lyra — COMMENTED** [review]
Writing style is too informal (blog post-y), and there are style guide issues (for inclusivity and grammar).

File: docs/src/main/sphinx/develop/connectors.rst [hunk 23]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+Predicate pushdown
+^^^^^^^^^^^^^^^^^^
+
+When executing a query with a ``WHERE`` clause, the query plan can
+contain a ``ScanFilterProject`` operation with a predicate constraint.
+
+A predicate constraint is a description of the constraint imposed on the
+results of the query as expressed in the ``WHERE`` clause. For example,
+``WHERE x > 5 AND y = 3`` translates into a ``Constraint`` where the
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:311]: just constraint, no initial cap and no double backticks

File: docs/src/main/sphinx/develop/connectors.rst [hunk 24]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+``summary`` field means the ``x`` column's domain must be greater than
+``5`` and the ``y`` column domain equals ``3``.
+
+When the query plan contains a ``ScanFilterProject`` operation, Trino
+tries to optimize the query by pushing down the predicate constraint
+into the connector by calling the ``applyFilter`` method of the
+connector metadata service. This method receives a table handle with
+all optimizations applied thus far, and should return either
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:319]: and returns either 

(no should - it is against developer style guide)

File: docs/src/main/sphinx/develop/connectors.rst [hunk 25]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+``Optional.empty()`` or a response with a new table handle derived from
+the old one.
+
+The query optimizer may call ``applyFilter`` for a single query multiple times,
+as it searches for an optimal query plan. It's important that connectors
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:324]: Connectors must return

File: docs/src/main/sphinx/develop/connectors.rst [hunk 26]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+return ``Optional.empty()`` from ``applyFilter`` if they can't apply the
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:325]: cannot

(no contractions - against developer style guide)

File: docs/src/main/sphinx/develop/connectors.rst [hunk 27]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+constraint for this invocation even if they support ``ScanFilterProject``
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:326]: invocation, even

File: docs/src/main/sphinx/develop/connectors.rst [hunk 28]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+pushdown in general. It's also important to return ``Optional.empty()`` if the
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:327]: Connectors must also return

File: docs/src/main/sphinx/develop/connectors.rst [hunk 29]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+* an expression for pushing down function calls,
+* a map of assignments from variables in the expression to columns,
+* (optional) a predicate which tests a map of columns and their values,
+  it cannot be held on to after the ``applyFilter`` call returns
+* (optional) a set of columns the predicate depends on, must be present
+  if predicate is present.
+
+A ``TupleDomain`` defines a mapping between columns and their domains.
+A ``Domain`` is either a list of possible values or a list of ranges.
+It also contains information about nullability.
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:343]: A constraint contains the following elements:

* A ``TupleDomain`` defining the mapping between columns and their domains. A ``Domain`` is either a list of possible values, or a list of ranges, and also contains information about nullability.
* Expression for pushing down function calls.
* Map of assignments from variables in the expression to columns.
* (optional) Predicate which tests a map of columns and their values;
  it cannot be held on to after the ``applyFilter`` call returns.
* (optional) Set of columns the predicate depends on; must be present
  if predicate is present.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 30]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+
+If both a predicate and a summary are available, the predicate is guaranteed to
+be more strict in filtering of values, and can provide a significant boost to
+query performance if used.
+
+However it's not possible to store a predicate in the table handle and use
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:349]: It is not possible to

File: docs/src/main/sphinx/develop/connectors.rst [hunk 31]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+and focus only on ``summary``.
+In such case, the summary can still be pushed down by storing it in the table
+handle.
+
+This overlap between the predicate and summary is due to historical reasons,
+as simple comparison pushdown was implemented first via summary, and more
+complex filters such as ``LIKE`` which required more expressive predicates
+were added later.
+
+If a constraint can only be pushed down partially, for example when a connector
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:362]: partially pushed down

File: docs/src/main/sphinx/develop/connectors.rst [hunk 32]
```diff
@@ -297,6 +298,324 @@ an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
...
+for a database that doesn't support range matching is used in a query with
```
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:363]: does not

**Review by m57lyra — COMMENTED** [review]
Looking better, but still some things to fix

**Review by m57lyra — APPROVED** [review]
LGTM!

**Comment by m57lyra** [comment]
@MiguelWeezardo - Did you actually make additional changes after my approval that i need to review or is this just the squash and merge?

**Comment by MiguelWeezardo** [comment]
> @MiguelWeezardo - Did you actually make additional changes after my approval that i need to review or is this just the squash and merge?

It's just a squash.

**Comment by MiguelWeezardo** [comment]
@m57lyra Should I involve a maintainer to get this merged? If so, I'll probably need an official OK from you.

**Comment by MiguelWeezardo** [comment]
@hashhar I'd appreciate if you took a look at this and possibly merge if you think this is useful.

**Comment by hashhar** [comment]
Sure, will take a look. Some quick comments follow.

**Comment by MiguelWeezardo** [comment]
@hashhar Could you take another look and merge if OK?

**Comment by MiguelWeezardo** [comment]
@hashar Could you take another look and merge if OK?

**Review by raunaqmorarka — APPROVED** [review]
